src/tests/unit_x509.cpp

```diff
@@ -1592,8 +1593,885 @@ Test::Result test_x509_tn_auth_list_extension_decode() {
+                  result.confirm("ipv4 entry 3 max", ranges[3].max().value() == ipv4_range_4.max().value(), true);
+               } else {
+                  result.confirm("ipv4 range has no entries", ranges.empty(), true);
+               }
+            } else {
+               result.confirm("ipv4 family inherit", choice.ranges().has_value(), false);
+            }
+         }
+
+         if(push_ipv6_family) {
+            auto family = dec_addr_blocks[dec_addr_blocks.size() - 1];
+            result.confirm("ipv6 family afi", ipv6_addr_family.afi() == family.afi(), true);
+            result.confirm("ipv6 family safi", ipv6_addr_family.safi() == family.safi(), true);
+            auto choice =
+               std::get<IPAddressBlocks::IPAddressChoice<IPAddressBlocks::Version::IPv6>>(family.addr_choice());
+            if(!inherit_ipv6) {
+               auto ranges = choice.ranges().value();
+               if(push_ipv6_ranges) {
+                  result.confirm("ipv6 entry 0 min", ranges[0].min().value() == ipv6_range_1.min().value(), true);
+                  result.confirm("ipv6 entry 0 max", ranges[0].max().value() == ipv6_range_1.max().value(), true);
+                  result.confirm("ipv6 entry 1 min", ranges[1].min().value() == ipv6_range_2.min().value(), true);
+                  result.confirm("ipv6 entry 1 max", ranges[1].max().value() == ipv6_range_2.max().value(), true);
+                  result.confirm("ipv6 entry 2 min", ranges[2].min().value() == ipv6_range_3.min().value(), true);
+                  result.confirm("ipv6 entry 2 max", ranges[2].max().value() == ipv6_range_3.max().value(), true);
+                  result.confirm("ipv6 entry 3 min", ranges[3].min().value() == ipv6_range_4.min().value(), true);
+                  result.confirm("ipv6 entry 3 max", ranges[3].max().value() == ipv6_range_4.max().value(), true);
+               } else {
+                  result.confirm("ipv6 range has no entries", ranges.empty(), true);
+               }
+            } else {
+               result.confirm("ipv6 family inherit", choice.ranges().has_value(), false);
+            }
+         }
+      }
+   }
+
+   result.end_timer();
+   #endif
+   return result;
+}
+
+Test::Result test_x509_ip_addr_blocks_extension_encode_edge_cases() {
+   Test::Result result("X509 IP Address Block encode edge cases");
+   #if defined(BOTAN_HAS_RSA)
+   result.start_timer();
+
+   using Botan::Cert_Extension::IPAddressBlocks;
+
+   auto rng = Test::new_rng(__func__);
+
+   const std::string sig_algo{"RSA"};
+   const std::string hash_fn{"SHA-256"};
+   const std::string padding_method{"EMSA3(SHA-256)"};
+
+   // trailing 0s, trailing 1s, and some arbitrary values
+   std::vector<uint8_t> edge_values = {2, 4, 8, 16, 32, 64, 128, 0, 1, 3, 7, 15, 31, 63, 127, 255, 12, 46, 123, 234};
+
+   for(size_t i = 0; i < edge_values.size(); i++) {
+      for(size_t j = 0; j < 18; j++) {
+         auto ca_key = make_a_private_key(sig_algo, *rng);
```

Comment:
Do you really need to create a new key for each case? I would think it sufficient to create one key for the entire test

Reply:
(Likewise for the self signed CA cert and the `X509_CA` instance)